rasa/core/actions/flows.py

```diff
@@ -73,3 +80,84 @@ async def run(
             events.append(ActiveLoop(None))
 
         return events
+
+
+class ActionCancelFlow(action.Action):
+    """Action which cancels a flow from the stack."""
+
+    def __init__(self) -> None:
+        """Creates a `ActionCancelFlow`."""
+        super().__init__()
+
+    def name(self) -> Text:
+        """Return the flow name."""
+        return ACTION_CANCEL_FLOW
+
+    async def run(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        metadata: Optional[Dict[Text, Any]] = None,
+    ) -> List[Event]:
+        """Cancel the flow."""
+        stack = FlowStack.from_tracker(tracker)
+        if stack.is_empty():
+            structlogger.warning("action.cancel_flow.no_active_flow", stack=stack)
+            return []
+
+        context = tracker.get_slot(FLOW_CONTEXT_SLOT) or {}
+        canceled_flow_frames = context.get("canceled_frames", [])
+
+        for frame_idx in canceled_flow_frames:
+            if frame_idx >= len(stack.frames):
+                structlogger.warning(
+                    "action.cancel_flow.frame_not_found",
+                    stack=stack,
+                    frame_idx=frame_idx,
+                )
+                continue
+            stack.frames[frame_idx].step_id = END_STEP
+
+        return [SlotSet(FLOW_STACK_SLOT, stack.as_dict())]
+
+
+class ActionCorrectFlowSlot(action.Action):
+    """Action which corrects a slots value in a flow."""
+
+    def __init__(self) -> None:
+        """Creates a `ActionCancelFlow`."""
+        super().__init__()
+
+    def name(self) -> Text:
+        """Return the flow name."""
+        return ACTION_CORRECT_FLOW_SLOT
+
+    async def run(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        metadata: Optional[Dict[Text, Any]] = None,
+    ) -> List[Event]:
+        """Correct the slots."""
+        stack = FlowStack.from_tracker(tracker)
+        if stack.is_empty():
+            structlogger.warning("action.correct_flow_slot.no_active_flow", stack=stack)
+            return []
+
+        context = tracker.get_slot(FLOW_CONTEXT_SLOT) or {}
```

Comment:
same as above ~ isn't the context part of the stack frame and not a global slot?

Reply:
yes it is, let's keep the discussion on the one comment below :+1: